datasetinsights/estimators/siam_mask.py

```diff
@@ -0,0 +1,2251 @@
+            )  # 15x15, 7x7
+            self.feature_size = (256 + 128) * block.expansion
+        else:
+            self.layer3 = lambda x: x  # identity
+
+        if layer4:
+            self.layer4 = self._make_layer(
+                block, 512, layers[3], stride=1, dilation=4
+            )  # 7x7, 3x3
+            self.feature_size = 512 * block.expansion
+        else:
+            self.layer4 = lambda x: x  # identity
+
+        for m in self.modules():
+            if isinstance(m, nn.Conv2d):
+                n = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
+                m.weight.data.normal_(0, math.sqrt(2.0 / n))
+            elif isinstance(m, nn.BatchNorm2d):
+                m.weight.data.fill_(1)
+                m.bias.data.zero_()
+
+    def _make_layer(self, block, planes, blocks, stride=1, dilation=1):
+        downsample = None
+        dd = dilation
+        if stride != 1 or self.inplanes != planes * block.expansion:
+            if stride == 1 and dilation == 1:
+                downsample = nn.Sequential(
+                    nn.Conv2d(
+                        self.inplanes,
+                        planes * block.expansion,
+                        kernel_size=1,
+                        stride=stride,
+                        bias=False,
+                    ),
+                    nn.BatchNorm2d(planes * block.expansion),
+                )
+            else:
+                if dilation > 1:
+                    dd = dilation // 2
+                    padding = dd
+                else:
+                    dd = 1
+                    padding = 0
+                downsample = nn.Sequential(
+                    nn.Conv2d(
+                        self.inplanes,
+                        planes * block.expansion,
+                        kernel_size=3,
+                        stride=stride,
+                        bias=False,
+                        padding=padding,
+                        dilation=dd,
+                    ),
+                    nn.BatchNorm2d(planes * block.expansion),
+                )
+
+        layers = []
+        layers.append(
+            block(self.inplanes, planes, stride, downsample, dilation=dd)
+        )
+        self.inplanes = planes * block.expansion
+        for i in range(1, blocks):
+            layers.append(block(self.inplanes, planes, dilation=dilation))
+
+        return nn.Sequential(*layers)
+
+    def forward(self, x):
+        x = self.conv1(x)
+        x = self.bn1(x)
+        x = self.relu(x)
+        # print x.size()
+        x = self.maxpool(x)
+        # print x.size()
+
+        p1 = self.layer1(x)
+        p2 = self.layer2(p1)
+        p3 = self.layer3(p2)
+        # p3 = torch.cat([p2, p3], 1)
+
+        # log_once("p3 {}".format(p3.size()))
+        p4 = self.layer4(p3)
+
+        return p2, p3, p4
+
+
+# Bottleneck class - part of Resnet
+class Bottleneck(nn.Module):
+    expansion = 4
```

Comment:
where should this live? is this a constant? a config hyperparameter? a property of the bottleneck itself?